controllers/agents/app/servicebinding_controller.go

```diff
@@ -47,13 +60,326 @@ type ServiceBindingReconciler struct {
+				c := &v1.Container{}
+				u := (*container).(map[string]interface{})
+				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u, c); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				if len(sb.Spec.Application.Containers) > 0 {
+					found := false
+					count := 0
+					for _, v := range sb.Spec.Application.Containers {
+						l.Info("container", "container value", v, "name", c.Name)
+						if v.StrVal == c.Name {
+							break
+						}
+						found = true
+						count++
+					}
+					if found && len(sb.Spec.Application.Containers) == count {
+						continue CONTAINERS_OUTER
+					}
+
+				}
+
+				for _, e := range sb.Spec.Env {
+					c.Env = append(c.Env, v1.EnvVar{
+						Name:  e.Name,
+						Value: string(psSecret.Data[e.Key]),
+					})
+
+				}
+				mountPath := ""
+				for _, e := range c.Env {
+					if e.Name == ServiceBindingRoot {
+						mountPath = path.Join(e.Value, mountPathDir)
+						break
+					}
+				}
 
-	// TODO(user): your logic here
+				if mountPath == "" {
+					mountPath = path.Join("/bindings", mountPathDir)
+					c.Env = append(c.Env, v1.EnvVar{
+						Name:  ServiceBindingRoot,
+						Value: "/bindings",
+					})
+				}
 
+				volumeMount := v1.VolumeMount{
+					Name:      volumeName,
+					MountPath: mountPath,
+					ReadOnly:  true,
+				}
+
+				volumeMountFound := false
+				for j, vm := range c.VolumeMounts {
+					if vm.Name == volumeName {
+						c.VolumeMounts[j] = volumeMount
+						volumeMountFound = true
+						break
+					}
+				}
+
+				if !volumeMountFound {
+					c.VolumeMounts = append(c.VolumeMounts, volumeMount)
+				}
+
+				nu, err := runtime.DefaultUnstructuredConverter.ToUnstructured(c)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				containers[i] = nu
+			}
+
+			l.Info("updated container with volume and volume mounts", "containers", containers)
+
+			l.Info("setting the updated containers into the application using the unstructured object")
+			if err := unstructured.SetNestedSlice(application.Object, containers, containersPath...); err != nil {
+				return ctrl.Result{}, err
+			}
+			l.Info("application object after setting the updated containers", "Application", application)
+		}
+
+		var conditionStatus metav1.ConditionStatus
+		var reason, state string
+
+		conditionStatus = "True"
+		reason = "Success"
+		state = "Ready"
+
+		l.Info("updating the application with updated volumes and volumeMounts")
+		if err := r.Update(ctx, &applications[index]); err != nil {
+			l.Error(err, "unable to update the application", "application", application)
+			conditionStatus = "False"
+			state = "Malformed"
+			reason = "failure"
+		}
+		l.Info("set the status of the service binding")
+		_, err = r.setStatus(ctx, psSecret.Name, sb, conditionStatus, reason, state)
+		if err = errors.Join(el, err); err != nil {
+			l.Info("error joining error logs")
+		}
+	}
+
+	if el != nil {
+		return ctrl.Result{}, el
+	}
 	return ctrl.Result{}, nil
```

Comment:
I think something is wrong here. To me, you should try to update the application and if you got an error you store it in a list of errors. Then if you got at least one error, you set status to `Failed`, otherwise you set `Success`.
Note also that the result of `errors.Join(param []error)` is the error that wraps all the errors you provided to the function.

I'm expecting something like this:

```
       errs := []error{}
       ...
       ... {
               ....
		l.Info("updating the application with updated volumes and volumeMounts")
		if err := r.Update(ctx, &applications[index]); err != nil {
			l.Error(err, "unable to update the application", "application", application)
                        errs = append(errs, err)
		}
	}
	if len(errs) != 0 {
	        cerr := errors.Join(el...)
                if err := setErrorStatus(cerr); err != nil {
                         return ctrl.Result{}, err
                }
                return ctrl.Result{}, cerr
        }

        if err := setSuccessStatus(); err != nil {
                return ctrl.Result{}, err
        }
        return ctrl.Result{}, nil
```

Reply:
I think the same logic has been used but written differently. The `conditionStatus`, 'state' and 'reason' are defined with positive values and if the update fails then we make them to false. By doing so instead of having two separate methods for setErrorStatus and setSuccessStatus we have a single method.
Good you pointed the errors.Join stuff, i would update that and equalize to el, for linting to succeed.

Reply:
Updated the controller logic to update the status of service binding only once, after we collect all the errors after application update.